### PR TITLE
Dtype for reduction

### LINKED ISSUE #357: FP16 reductions are returning Float instead of FP16
## 🐛 Bug

The issue is our reduction kernels are not returning FP16 data when they consume FP16 input.  The example below is an inner dimension reduction.

```
__global__ void kernel1(Tensor<__half, 2> T0, Tensor<float, 1> T3){                                                                                               
  alignas(4) extern __shared__ char array[];                                                                                                                      
  void* shared_mem = array;
  if ( ( ( ( blockIdx.x * 16 ) + threadIdx.y ) < T0.size[0] ) ) {
    T3[ ( ( ( blockIdx.x * 16 ) + threadIdx.y ) * T3.stride[0] ) ]                                                                                                
       = float(0);                                                                                                                                                
  }
  float T4[1];
  if ( ( ( ( blockIdx.x * 16 ) + threadIdx.y ) < T0.size[0] ) ) {                                                                                                 
    T4[ 0 ]
       = float(0);                                                                                                                                                
  }
  for(size_t i10 = 0; i10 < ( ceilDiv(( ceilDiv(T0.size[1], 4) ), 32) ); ++i10 ) {                                                                                
    float T2[4];                                                                                                                                                  
    float T1[4];
    if ( ( ( ( ( blockIdx.x * 16 ) + threadIdx.y ) < T0.size[0] ) && ( ( ( ( ( i10 * 32 ) + threadIdx.x ) * 4 ) + ( 4 - 1 ) ) < T0.size[1] ) ) ) {                
      for(size_t i12 = 0; i12 < 4; ++i12 ) {                                                                                                                      
        T1[ i12 ]
           = __half2float(T0[ ( ( ( blockIdx.x * 16 ) + threadIdx.y ) * T0.stride[0] ) + ( ( ( ( i10 * 32 ) + threadIdx.x ) * 4 ) + i12 ) ]);                     
        T2[ i12 ]
           = T1[ i12 ]
           + float(0.0010000000474974513);                                                                                                                        
        T4[ 0 ]
           = T4[ 0 ]
           + T2[ i12 ];                                                                                                                                           
      }
    } else {
      for(size_t i12 = 0; i12 < 4; ++i12 ) {
        if ( ( ( ( ( blockIdx.x * 16 ) + threadIdx.y ) < T0.size[0] ) && ( ( ( ( ( i10 * 32 ) + threadIdx.x ) * 4 ) + i12 ) < T0.size[1] ) ) ) {                  
          T1[ i12 ]
             = __half2float(T0[ ( ( ( blockIdx.x * 16 ) + threadIdx.y ) * T0.stride[0] ) + ( ( ( ( i10 * 32 ) + threadIdx.x ) * 4 ) + i12 ) ]);                   
        }
        if ( ( ( ( ( blockIdx.x * 16 ) + threadIdx.y ) < T0.size[0] ) && ( ( ( ( ( i10 * 32 ) + threadIdx.x ) * 4 ) + i12 ) < T0.size[1] ) ) ) {                  
          T2[ i12 ]
             = T1[ i12 ]
             + float(0.0010000000474974513);                                                                                                                      
        }
        if ( ( ( ( ( blockIdx.x * 16 ) + threadIdx.y ) < T0.size[0] ) && ( ( ( ( ( i10 * 32 ) + threadIdx.x ) * 4 ) + i12 ) < T0.size[1] ) ) ) {                  
          T4[ 0 ]
             = T4[ 0 ]
             + T2[ i12 ];                                                                                                                                         
        }                                                                                                                                                         
      }                                                                                                                                                           
    }                                                                                                                                                             
  }
  if ( ( ( ( blockIdx.x * 16 ) + threadIdx.y ) < T0.size[0] ) ) {
    blockReduce< true, false, false > ( T3[ ( ( ( blockIdx.x * 16 ) + threadIdx.y ) * T3.stride[0] ) ], T4[ 0 ], reduction_add_float, threadIdx, blockDim, static_cast<float*>(shared_mem));
  }                                                                                                                                                               
}                                                                                                                                                                 
} 
```                                                                                      
## To Reproduce

You can use the code in this PR:

```
https://github.com/csarofeen/pytorch/pull/352
```

This is the command to run:
```
python -m tensorexpr --print-kernel --device gpu --dtype float16 --mode fwd --jit_mode trace --cuda_fuser=nvf --input-iter 640:640:640,1024:1024:pow2 reduce2d_inner
```

### PR DESCRIPTION
**Pull request by jjsjann123**
Fixes #357

Two things in this PR:
1. Do type propagation even for profiling executor -> this is the root cause for bug reported in #357
2. Allow dtype argument in `sum`, which is simply handled in our type propagation.

### CODE REVIEW

**Comment by jjsjann123** [comment]
I'm hitting a kernel issue with the support. fp16 kernel is giving me wrong output, looks like predicate is wrong. I'll open another issue to track it.

File: torch/csrc/jit/codegen/cuda/shape_inference.cpp [hunk 1]
```diff
@@ -148,7 +148,14 @@ class NaiveTypePropagator {
         break;
       }
       case aten::sum: {
-        const auto out_type = node->input(0)->type()->cast<TensorType>();
+        auto out_type = node->input(0)->type()->cast<TensorType>();
+
+        if (!node->input(3)->type()->isSubtypeOf(
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/shape_inference.cpp:154]: add a comment explaining what's going on?

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 2]
```diff
@@ -507,6 +507,14 @@ class IrParser {
             // we don't support cast of output types yet;
             if (!node->inputs()[3]->type()->isSubtypeOf(
                     static_cast<c10::TypePtr>(NoneType::get()))) {
+              // We can only handle output as half and float;
+              if (auto opt_ivalue = toIValue(node->input(3))) {
+                auto scalar_type = opt_ivalue->toScalarType();
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: const?

File: torch/csrc/jit/codegen/cuda/manager.cpp [hunk 3]
```diff
@@ -224,18 +224,7 @@ void compileCudaFusionGroup(Node* fusion_node) {
...
-    //       at critical path. We take the trade-off here as profiling executor
-    //       is the future;
-    //
-    // Type propagation that's here just to cover corner case, incase type
-    // propagation failed in the original subgraph. We currently need output
-    // types in order to support fp16, where we cast input to fp32 and output
-    // back to fp16.
-    TypePropagate(graph);
-  }
+  TypePropagate(graph);
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/manager.cpp:232]: is this still related to the corner case described in the original comment? if that's the case maybe we should preserve the comment?

**Review by tlemo — APPROVED** [review]

**Pull request merged**